fix: OPTIC-1794: Improve handling of missing tasks and projects when fetching data within DataManager

bmartel opened the pull request:
This pull request includes a couple updates to error handling within the `DataManager` component. We are introducing more graceful ways of handling missing tasks and project data, siding with redirects and more subtle toast messages instead of crashing the page and leaving the user in an unusable state of the application. We are now no longer going to treat polling data the same as normal data requests, as one is non critical and nice to have, the other must exist for the application to work. In the case of polling `interaction="timer"` data we will ignore these errors completely and allow it to retry. The non polling data fetching the main project or task will be handled with a redirect back to the page previous, DataManager of the project (missing task), or the projects list itself (missing project) in that order based on what the missing data type is.

[Kapture 2025-03-11 at 20.40.27.webm](https://github.com/user-attachments/assets/d1080771-ec36-41d5-a4ae-d8c9dca276ab)


### Error Handling and Redirection Improvements:

* [`web/apps/labelstudio/src/pages/DataManager/DataManager.jsx`](diffhunk://#diff-ee4ec91c0e13fd4685594a907d22c9ccae9ef8cce012fd603d19584690cb124aL99-R107): Enhanced the crash event handling to redirect users based on the error details, either to the home page or the projects page.
* [`web/libs/datamanager/src/stores/AppStore.js`](diffhunk://#diff-d1e11a839827428aa6aada8e027a22b057a463ccb73c693e464c6e0784a9825fL742-R758): Updated the `crash` method to accept an options parameter, allowing conditional redirection based on the error context.
* [`web/libs/datamanager/src/stores/AppStore.js`](diffhunk://#diff-d1e11a839827428aa6aada8e027a22b057a463ccb73c693e464c6e0784a9825fL507-R516): Modified the project fetching logic to handle cases where the project does not exist, triggering a crash with redirection.

### Project and Task Data Handling:

* [`web/libs/datamanager/src/stores/AppStore.js`](diffhunk://#diff-d1e11a839827428aa6aada8e027a22b057a463ccb73c693e464c6e0784a9825fR487-R498): Added a check for the length of the new project data to ensure valid comparison and updating of the project state.
* [`web/libs/datamanager/src/stores/DataStores/tasks.js`](diffhunk://#diff-524bd37e2870a2414a48568a8f21257ae07d41ce3e1bfe8fecbf02a0ddf4ce5bL148-R160): Enhanced the task data retrieval process to handle task cancellations and non-existent tasks, invoking a crash with redirection if necessary.

bmartel committed:
feat: OPTIC-1746: Improve global error message handling by showing toast messages in favour of modals and increased Sentry reporting

bmartel committed:
linting

bmartel committed:
Capture error boundary in sentry, remove stacktrace from Modal

bmartel committed:
linting

bmartel committed:
don't throw an error in test from useToast as it forces every consumer of the hook to mock it in order to stop it from causing errors even if unused

bmartel committed:
Update web/apps/labelstudio/src/components/Modal/Modal.jsx

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-optic-1746'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13697181579

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13697181579

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-optic-1746

bmartel committed:
adding better 404 handling

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13716135763

bmartel committed:
only automatically handle 404 errors if not inline

bmartel committed:
linting

bmartel committed:
fix errorFilter checking resulting in implementation breakages with empty result error key

bmartel committed:
fix: OPTIC-1794: Improve handling of missing tasks and projects when fetching data within DataManager

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-optic-1746'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13783174190

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13783174190

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-optic-1794

bmartel committed:
Merge remote-tracking branch 'origin/fb-optic-1746' into fb-optic-1794

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-optic-1794

bmartel committed:
Make the messaging more streamlined

bmartel committed:
linting

hlomzik commented on web/apps/labelstudio/src/pages/DataManager/DataManager.jsx:
```diff
@@ -96,7 +96,27 @@ export const DataManagerPage = ({ ...props }) => {
+      }
+
+      if (isMissingTaskError) {
+        history.push(buildLink("/", { id: params.id }));
```
Current code will redirect to pages like `/projects/123/` with trailing slash which we normally don't use
```suggestion
        history.push(buildLink("", { id: params.id }));
```

hlomzik commented on web/apps/labelstudio/src/pages/DataManager/DataManager.jsx:
```diff
@@ -96,7 +96,27 @@ export const DataManagerPage = ({ ...props }) => {
+      if (isMissingTaskError || isMissingProjectError) {
+        toast.show({
+          message: `The ${
+            isMissingTaskError ? "task" : "project"
```
I'd move this to variable before the `toast.show()` for readability

hlomzik commented on web/libs/datamanager/src/stores/AppStore.js:
```diff
@@ -484,17 +484,18 @@ export const AppStore = types
       try {
         const newProject = yield self.apiCall("project", params);
         const projectLength = Object.entries(self.project ?? {}).length;
+        const newProjectLength = Object.entries(newProject ?? {}).length;
```
The naming is really confusing here. better to rename it to something like `isProjectEmpty` and store boolean directly. We don't actually need a number of properties in project.

bmartel replied:
```suggestion
        const hasExistingProjectData = Object.entries(self.project ?? {}).length > 0;
        const hasNewProjectData = Object.entries(newProject ?? {}).length > 0;
```

hlomzik reviewed (approved):
Some non-critical observations but would be really great to fix them, they are simple and precise.

bmartel committed:
Update web/apps/labelstudio/src/pages/DataManager/DataManager.jsx

Co-authored-by: Andrew <hlomzik@gmail.com>

bmartel commented on web/libs/datamanager/src/stores/AppStore.js:
```diff
@@ -484,17 +484,18 @@ export const AppStore = types
 
         self.needsDataFetch =
-          options.force !== true && projectLength > 0
+          options.force !== true && projectLength > 0 && newProjectLength > 0
```
```suggestion
          options.force !== true && hasExistingProjectData && hasNewProjectData
```

bmartel committed:
Apply suggestions from code review

bmartel committed:
applying feedback from CR

AndrejOros commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13830987940)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-optic-1794'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13830987940

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13830987940

bmartel committed:
don't cancel, this apparently causes a regression with regards to preloaded tasks

AndrejOros commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13834430241)

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13834430241
